_posts/2020-10-23-fine-grained-access-control-for-search.md

```diff
@@ -0,0 +1,99 @@
+
+At this point you have a user, `public-item-user` and a role `public-item-viewer` but they aren’t connected. In Open Distro this is called *mapping*. To map a role to a user, go back to “Security” and then select “Roles” and find our user (`public-item-viewer`) and click on it (you may need to use the search box). It should bring you to `<yourhost>:<yourport>/app/opendistro_security#/roles/view/public-item-viewer`.  Now go to the tab “Mapped Users” - the list will be empty so click “Map users." At this point you’ll be at the URL `<yourhost>:<yourport>/app/opendistro_security#/roles/edit/public-item-viewer/mapuser`.  Under the heading "Internal users" go to the "Internal users" field and select `public-item-user`. Then click "Map."
+
+Now your user (`public-item-user`) has all the powers provided by the role `public-item-viewer`. Let’s test it out by going to Kibana in a private or incognito window. This trick will allow you to be logged into two users at the same time. Go to Dev Tools (<yourhost>:<yourport>/app/dev_tools#/console) and enter the following query:
+
+```
+GET /ecommerce-items/_search
+{
+  "query": {
+    "match" : {
+      "title": "brown"
+    }
+  }
+}
+```
+
+The result will show:
+
+```
+{
+    ...,
+    "hits" : [
+        {
+        "_index" : "ecommerce-items",
+        "_type" : "_doc",
+        "_id" : "brown-mug",
+        "_score" : 0.13353139,
+        "_source" : {
+            "title" : "Brown Mug",
+                "description" : "This mug features a large handle and a marbled, brown ceramic design.",
+                "price" : 5.99
+            }
+        }
+    ]
+}
+```
+
+Notice that the `private` object is gone. Our user and role has filtered this data right out.
+
+One additional note here - we setup these two roles to *only* have access to this particular index. That means that if,  by some mistake or happenstance the credentials were released for these users or an application bug somehow allowed for passing in other indices, you’re covered. No other index managed by Open Distro would be at risk of disclosure as access control would deny the query. You can think of this as an implementation of the principle of least privilege.
+
+To setup a route for the employee section of your application, you would create a role called `private-item-viewer`  that is allowed to see all information. Repeat the same steps as above except skipping the step where you exclude `private` in the “Field level security section”. Then you would create a user called `private-item-user` with the role of `private-item-viewer`. All of the requests from to backend for employees would uses this user to make the requests, enabling them to have access to all the fields in the object `private`.
```

Comment:
I think you can convert this to present tense and just make the whole thing read a bit more cleanly.

To set up a route for the employee section of your application, create a role called `private-item-viewer` that has permissions to see all information. Repeat the process above, except skip the step where you exclude the `private` field. Then create a user...

Etc. etc.